### ISSUE TITLE
Sync silently ignores new files with old timestamps

### ISSUE THREAD

**Issue by jacquesh** [laurent22/joplin#6517]
When syncing to the file system, new files with a timestamp older than the last sync timestamp get silently ignored (instead of being pulled into Joplin).

This situation arises for me because I run Joplin on my PC and phone, have them both sync to the filesystem and then use [syncthing](https://syncthing.net/) to sync the two. syncthing maintains file last-modified metadata across devices, meaning that if my mobile device is not busy syncing (could be that syncing is off, I'm out of network range, off wifi, etc) and I add a note, wait a day and then let it sync, then the file on my PC will have a creation date of right now and a last-modified date that is a day in the past. Joplin ignores it.

## Environment

Joplin version: 2.7.15 (Although this has been happening for at least a year, I updated recently and I *think* I was on 2.3.6 before that)
Platform: Desktop, Windows 10 x64 21H2/19044

## Steps to reproduce

1. Set joplin to sync to the filesystem and ensure it is all synced up.
2. Close Joplin
3. Create a new note file (I copied an existing one and changed the name & ID in the file to a new hash)
4. Set the file's last-modified time to before your last Joplin sync
    * e.g in powershell: `(Get-Item "c6989cfc4d4d9114591431119fe5e8a5.md").LastWriteTime=("15 May 2022 19:38:00")`
5. Open Joplin & sync
6. See that it does not pull in the new file

## Thoughts following my own short investigation into the synchronisation code

I've looked through the sync code in Joplin and I see it is relying entirely on the last-modified time of the file. In particular it will ignore any file with a timestamp older than Joplin's last sync time *even if the file is unknown to Joplin*. It seems to me that Joplin should instead read in new files even if they're older than the last sync time but there may well be other factors that I'm not aware of. I know little about the filesystem format but at the very least Joplin would need to check the contents of each new file to know what its reading in (my local instance has 200 notes but >500 files, some of which appear to be recordings of edits? Presumably these contains data from tables in the sqlite DB other than the "notes" table?)

Admittedly the same problem can happen with file edits. This is more difficult to resolve since you can't use the existence of the file as proof that a sync is needed. In that case, you might be able to use the file-specific sync time (in the sync_items table, assuming that actually contains what the name suggests) to do a somewhat better job, but short of hashing the file contents (which would be very robust but also far slower than checking file times and obviously problematic on all of the non-filesystem sync targets, possibly worth doing only for the filesystem sync target?) I'm not sure there's a great solution for that case. If the conclusion is that this behaviour is a problem then maybe solving the new-file case is at least a good start?

I can see about changing my non-joplin file-sync setup so that the file timestamps are updated when the file syncs (which would prevent this issue from happening to me in the first place) but that might not be as easy as one might hope and would only solve the problem with my particular setup (other file-system setups may run into the same issue).

This could possibly be the related to #5099 and #6346

**Comment by tomasz1986** [comment]
> This situation arises for me because I run Joplin on my PC and phone, have them both sync to the filesystem and then use [syncthing](https://syncthing.net/) to sync the two. syncthing maintains file last-modified metadata across devices, meaning that if my mobile device is not busy syncing (could be that syncing is off, I'm out of network range, off wifi, etc) and I add a note, wait a day and then let it sync, then the file on my PC will have a creation date of right now and a last-modified date that is a day in the past. Joplin ignores it.

I've experienced the exact same problem, also when using Syncthing. In my case, I believe this usually happens when Joplin performs synchronisation while Syncthing is still downloading the files, which results in a partial Joplin sync. Then, when trying to sync again, Joplin doesn't detect the notes that have been downloaded in between. I wasn't sure what the problem was about exactly though. Thank you so much for finding the actual culprit.

Is there any workaround that you'd suggest to do at the user's side as a temporary solution? Other than disabling automatic sync completely and relying on manual sync only, of course 😉. When encountering the error myself, I've up to now relied on the "delete local data and re-download from sync target" button, but this requires a manual action, while I'd prefer to rely on an automated solution here.

I think https://discourse.joplinapp.org/t/sync-issues-using-filesystem-sync-and-sync-tool/2652 from 3 years ago (!) talks about the same issue. It mentions touching all files in order to update their timestamps as a possible workaround. Is there any danger in doing just that (e.g. with a periodically run script)?

**Edit 1:**

I performed the whole touch thing, and after that Joplin was able to discover a dozen items that had been missing, which I wasn't even aware of… I haven't observed any peculiarities after doing it. However, I must mention that the process is slow, especially on low-end hardware and mobile devices, depending on the number of notes.

Just to be thorough, and also because people in that forum thread were asking about how to do touch in Windows, the actual command is `copy /b file.md+` (or `for /f %a in ('dir /b *.md') do copy /b %a+` to process all the md files).

**Edit 2:**

I think a possible workaround could to be sync the folder in Syncthing to a different path first, and then use some kind of a background script that would make sure that the folder has been synced completely (e.g. if no changes have happened in the last x minutes), and only then copy that folder in a single pass to yet another folder that would be used strictly as the sync target for Joplin. Doing so would eliminate the partial sync issue. Obviously, this is possible only on the desktop, and not with the mobile versions of Joplin, which likely suffer from the same problem. Also, it wouldn't be perfect either, as the synchronisation in Syncthing can still be interrupted, e.g. by a disconnect, and then the sync state will end up being incomplete anyway.

**Edit 3:**

I think I've managed to hack up a PowerShell script that "touches" only newly added files, while skipping everything else.

```powershell
Get-ChildItem -Path *.md | Where-Object {$_.LastWriteTime -le (Get-Date)} | ForEach-Object {$_.LastWriteTime = (Get-Date).AddYears(1)}
```

The script checks the modified time of all notes whether it's equal or older than the current moment, and then, if the condition matches, it updates it to 1 year from now. On the first run, this forces Joplin to rescan all notes, but later on, it only makes it rescan new/modified notes, while skipping those that have had their date modified already.

I think the date can be set even further into the future, so that even once the 1-year period has passed, Joplin will still not be forced to go through those notes, but the value of "1" year just seemed like a safe bet.

**Comment by trymeouteh** [comment]
This has been a big issue for me when using the file system + syncthing to sync notes. I been using file system + syncthing to sync my notes for 4 months now and every month I end up having this issue and having backup my notes and reset the Joplin apps on all of my devices and redo the sync. It is very irritating.

I would like to see Joplin fix this as this is considered a major bug. I like Joplin but lately it has not been performing too well, I hope Joplin get some contributors to help touch up the app since it is a good app overall and would like to see it perform to its fullest potential.

**Comment by Szybet** [comment]
thanks @tomasz1986 for the investigation, dzięki

I have created a linux version of his command
```
find -mtime -5 -exec touch {} +
```
Every file older than 5 days will be re-synced so it will be showed.
Also with syncthing, detecting conflicts doesn't work, but the files conflicter are still there so:
```
find . -type f -name '*conflict*' -exec kate {} +
```

**Comment by XilinJia** [comment]
Perhaps something related to "Edit 3:" from @tomasz1986 can be implemented in Joplin, as follows:

Instead of touching the files, Joplin can set back the timestamp of the previous scan and so at the next scan (either manual or automatic) it will scan based on the reset timestamp.  How to set back the timestamp?  I would suggest to provide a UI to the user so that, once he discovered some abnormalities with the sync, he set back the timestamp to say one week ago, or one month ago, etc.  Would that an easy fix?

**Comment by individual-it** [comment]
@Szybet I think the command `find -mtime -5 -exec touch {} +` will find and touch all files that was modified within the last 5 days.
from find man page: `File's data was last modified less than, more than or exactly n*24 hours ago`
To fix this problem, don't you want to touch all files that haven't been touched recently? For that you would need `find -mtime +5 -exec touch {} +`
Or do I misunderstand something?

**Comment by Szybet** [comment]
```
A numeric argument n can be specified to tests (like -amin,
-mtime, -gid, -inum, -links, -size, -uid and -used) as

+n     for greater than n,

-n     for less than n,

n      for exactly n.
```
yes, you misunderstood

**Comment by laurent22** [comment]
>  I add a note, wait a day and then let it sync, then the file on my PC will have a creation date of right now and a last-modified date that is a day in the past.

On Joplin side, wouldn't the solution be to look at both the creation and mast-modified timestamps and use whatever is the latest one for comparison?

Would that also address the case where SyncThing and Joplin are synchronising at the same time?

**Comment by tomasz1986** [comment]
Does Android even support file creation time? I don't think it does (e.g. see https://android.stackexchange.com/questions/232633/android-11-supports-creation-time-crtime-btime-for-local-folders-and-files).

> Would that also address the case where SyncThing and Joplin are synchronising at the same time?

I think it's a bit more than just syncing at the same time. For example:

1. Start sync with Syncthing.
2. The connection drops, leading to partial sync.
3. Joplin syncs its database.
4. Syncthing restores the connection later, and syncs the rest of the files.
5. The newly added files, however, are now completely ignored by Joplin.

I believe this is also why this bug/issue is basically bound to happen at some point.

**Comment by laurent22** [comment]
Hmm, but the current logic seems sound to me: Joplin keeps the latest timestamps of the files that have been synced. So let's say we start with these file timestamps:

- t0
- t1
- t2

Joplin will take the list of current files and process this - so it will process the file with timestamp t0, then t1, then t2.

If SyncThing is still synchronising in the background too, any new file it will add will have (or should have) a timestamp that comes after t2 - say t3, then t4.

Once Joplin is done synchronising, it's going to save the timestamp t2. Then next time it synchronises, it will start from there and look at t3, then t4.

For me, this should work provided timestamps are making sense - i.e. if I add a file to a folder, updated and created time should be the current time, not a time in the past. Or if SyncThing is doing something fancy and changing timestamps, at least one of the timestamps should be correct, which is why I suggested looking at both created and updated time. 

> The connection drops, leading to partial sync.

This should be fine - if Syncthing synchronises later, logically the created files should take the current time, not one in the past. Since it's a new file, it shouldn't have a timestamp in the past.

**Comment by tomasz1986** [comment]
I think what happens is basically something like this:

1. Create a few notes on the desktop. Let's say their modified time is 6:00.
2. Syncthing manages to synchronise only some of the notes, then it disconnects.
3. Joplin on Android performs its sync at 6:30, adding the new notes into the database.
4. Syncthing reconnects at 7:00 and syncs the remaining notes.
5. The newly synced notes still have their "modified time" at 6:00, but Joplin's last sync was at 6:30, so it ignores everything before that time.

There is no "created time" on Android, only "modified time", and that is synced with the files.

I work around the problem by using the PowerShell script to set the modified time to the future.

I think someone could try to reproduce the problem simply by manually (i.e. without using Syncthing at all) copying some of the MD files first, syncing those with Joplin, then copying the rest and trying to sync again.

**Comment by SiddharthManthan** [comment]
This issue makes file system sync unusable. Perhaps a fix could be priortized ?

**Comment by Kamul-PL** [comment]
Has anyone managed to find any sensible (safe and convenient) solution to this problem?

**Comment by aflip** [comment]
I'm syncing between two linux laptops and i do not face this issue. I manually edited the update time and create time of notes using the GUI, synced, and the notes show up just fine. This seems like it was solved.

**Comment by tomasz1986** [comment]
> I'm syncing between two linux laptops and i do not face this issue. I manually edited the update time and create time of notes using the GUI, synced, and the notes show up just fine. This seems like it was solved.

The issue isn't about the time inside Joplin, it's about the modified time of the actual MD files on the disk.

**Comment by Aman-1412** [comment]
\<snip\>
May have been wrong about a few things here. Deleted the comment.

**Comment by heitmanr** [comment]
some workarounds to "kick" some kind of re-sync missing notes are in this thread.

But how to be sure, that sync is now 100% done? some "CSV" with a list of all notes, maybe 3 rows
1) Folder
2) Note-Title
3) last modification-time
would be great to allow some diff between multiple computers...

..unfortunately "Note Overview"-plugin is shown as "incompatible" [no reason why mentioned]... any idea?

**Comment by heitmanr** [comment]
> Hmm, but the current logic seems sound to me: Joplin keeps the latest timestamps of the files that have been synced. So let's say we start with these file timestamps:
> 
> * t0
> * t1
> * t2
> 
> If SyncThing is still synchronising in the background too, any new file it will add will have (or should have) a timestamp that comes after t2 - say t3, then t4.
> 
no - is this assumption maybe the root-cause of the issue?

File-System-Sync Tools not neccessarily sync everything sequentialy - the filesystem-sync-tool could
a) first sync a file with "t0"
b) than sync a file with "t2"
...joplin might sync now
c) than sync a file with "t1"
...joplin will lose this note

What about offering an option for a manual or scheduled "daily?" ... "full resync"?

Why not tracking a list of all processed ...MD-Files to have a change to find "overlooked" files?

**Comment by individual-it** [comment]
syncthing actually has a setting for the sync order, would it solve the issue to set it to `Oldest First`?

![Image](https://github.com/user-attachments/assets/7583e694-ab3e-4243-9048-d1f5dcfedf6b)

**Comment by tomasz1986** [comment]
> syncthing actually has a setting for the sync order, would it solve the issue to set it to `Oldest First`?
> 
> ![Image](https://github.com/user-attachments/assets/7583e694-ab3e-4243-9048-d1f5dcfedf6b)

I don't think this will make any difference, because the modified time of those files will still be older than the Joplin's last sync time (if Joplin performs its sync in the middle of Syncthing's file sync), i.e. the files synced later will still end up being ignored by Joplin.

**Comment by heitmanr** [comment]
IMHO Joplin should sync reliably with any Filesynch Strategy.

If not, it should warn users when they choose Filesystem-Sync that they will lose data 

...If we have a look at the forum or the issue tracker, the question is not "if" but just "when" loss will occur.

**Comment by tomasz1986** [comment]
I think the only bulletproof way to fix these kind of problems would be not to rely on timestamps at all (which I believe to be a good idea anyway, as timestamps often fluctuate depending on the OS, e.g. Android, or the filesystem, e.g. FAT, which makes them unreliable).

For the record, Syncthing itself doesn't care about timestamps. It tracks actual file modifications instead to determine which version is newer, older, or same.

**Comment by jpvg10** [comment]
> if I add a file to a folder, updated and created time should be the current time, not a time in the past

This is not true at all. Take any file on your Windows PC (same thing on Linux as far as I could check) and copy-paste it to another folder. The creation date is right now, but the modification date is whatever the modification date of the original file was (which could be years ago).

This was already identified by the OP as the root of the problem 3 years ago. You can't ignore files with old modification dates because it's totally possible to have a "new" file with an old modification date.

**Comment by jpvg10** [comment]
> For the record, Syncthing itself doesn't care about timestamps. It tracks actual file modifications instead to determine which version is newer, older, or same.

I think this might be wrong according to [the Syncthing docs](https://docs.syncthing.net/users/syncing.html#scanning)?

> During a rescan (regardless whether full or from watcher) the existing files are checked for changes to their **modification time**, size or permission bits.

But anyways, I think Syncthing in this case is not the culprit, but Joplin by ignoring files.

**Comment by danielb2** [comment]
--> bump <--

**Comment by tomasz1986** [comment]
> > For the record, Syncthing itself doesn't care about timestamps. It tracks actual file modifications instead to determine which version is newer, older, or same.
> 
> I think this might be wrong according to [the Syncthing docs](https://docs.syncthing.net/users/syncing.html#scanning)?
> 
> > During a rescan (regardless whether full or from watcher) the existing files are checked for changes to their **modification time**, size or permission bits.

Just for clarification, Syncthing checks timestamp to know whether the file has changed or not, but it doesn't rely on it to determine which version is newer. For example, there are legitimate cases where you want to restore an older version of a file, and despite its older timestamp Syncthing will treat the restored file as the newest version. Timestamps themselves aren't reliable enough to use them as an indicator of which file is newer, because they can easily be impacted by factors like the filesystem (e.g. FAT32 with its imprecise timestamps) or timezone changes (and especially shifts due to daylight saving time).

**Comment by jpvg10** [comment]
Yes, I agree timestamps are unreliable. That's why Joplin shouldn't be using them to ignore files on sync...

**Comment by danielb2** [comment]
wouldn't it be pretty easy to implement a force of a whole new scan which ignores the timestamps and checks every file? seems lke this would solve the problem for the one-offs.

**Comment by mrjo118** [comment]
> wouldn't it be pretty easy to implement a force of a whole new scan which ignores the timestamps and checks every file? seems lke this would solve the problem for the one-offs.

Doing a full rescan might be a bit overkill, as it would be slow if you have a lot of objects in your profile, especially because newer versions of Android have performance issues when using file system sync (see https://discourse.joplinapp.org/t/sync-is-blocked-while-joplin-on-android-slowly-deletes-remote-items/45130/22).

Rather than implementing a full re-scan, maybe a 're-sync this note' option could be added to Joplin desktop and mobile in the context menu of a note, which will re-sync the current note only, but ignore the last sync time when doing so. This could help mitigate the consequences of the issue, by allowing you to resync a note where you notice it has not synced properly.

Depending on the situation, this option would do one of 3 things, where the updated_time within the MD file of the note (instead of the modified time on the file) is used to determine whether a note is an old or new version:
1. Do nothing, as it is up to date
2. Download a newer change that was skipped
3. Upload the current note, if the current change is newer than the version on remote

It may be easier to use the existing sync mechanism to re-sync the note, by just tinkering with the sync time for the note in the db, in which case it would be a mark for re-sync option instead. But I'm not sure if that would work for scenario 3, and it may be more reliable to have a separate re-sync function, which has it's own step in the sync process.

Also, linking my comment on the other issue for reference:
https://github.com/laurent22/joplin/issues/12706#issuecomment-3089514464

**Comment by danielb2** [comment]
> Doing a full rescan might be a bit overkill, as it would be slow if you have a lot of objects in your profile, especially because newer versions of Android have performance issues when using file system sync

I want to note that I said scan and not sync. It has to scan for new timestamp changes anyway, but it's the date in the files that matter. I don't know the limitations of the performance issue.

**Comment by mrjo118** [comment]
> > Doing a full rescan might be a bit overkill, as it would be slow if you have a lot of objects in your profile, especially because newer versions of Android have performance issues when using file system sync
> 
> I want to note that I said scan and not sync. It has to scan for new timestamp changes anyway, but it's the date in the files that matter. I don't know the limitations of the performance issue.

When using file system sync in Joplin, a rescan to check for any missed changes would effectively be the same thing as a resync (Joplin has no knowledge about Syncthing, file system sync will sync data to a directory outside of the Joplin sandboxed storage on Android). Fetching any and every item on Android, in order to check for new timestamp changes is going to be affected by the performance issue.

I discussed the issue further with @tomasz1986 outside of the forum thread, and concluded that the problem is basically a performance limitation in newer versions of Android when using SAF to access the file system (this affects file system sync only, other sync methods are unaffected). This can't be avoided unless either the directory stays within the sandboxed storage for Joplin (then Syncthing can't access it), or the Joplin app applies with Google to get an access all files permission added to the app (which would surely get rejected, because it is intended for file manager apps)

**Comment by mrjo118** [comment]
FYI there is already a 'Delete local data and re-download from sync target' option in the advanced section of Synchronize settings on Joplin desktop. That kind of is a re-scan, except it deletes and re-downloads every item instead of just updates each item that needs an update. I don't think update vs recreation makes much difference in performance really, as the heaviest part is the IO operations on the file system, which is likely the same in both cases.

I think the 'Delete local data and re-download from sync target' logic could be used as a basis a re-sync this note feature, but just isolated to a single note.

**Comment by mrjo118** [comment]
Going back to the issue detailed on this thread, I thought the issue was with the sync_time on the sync_items records which correspond to the notes. But @tomasz1986 said the issue also affects new items synced on one device which do not get received on another device (so a re-sync this note feature won't help if the note does not exist locally), so it seems like the issue might be fully due to the logic for the basic delta implementation (file-api.ts basicDelta function).

From what I can tell from a brief revisit of the basic delta, there seems to be a global last sync time which is cached, separate to the sync time on each individual item. It looks like the sync_time of each item is just used for determining note conflicts, not whether a new version of the note needs to be downloaded.

If the issue is with basic delta only, this may be easier to solve or at least mitigate. I'm wondering if the issue is just with Syncthing syncing within just a second or 2 of a file being synced by Joplin, in which case you may be able to easily solve the problem by subtracting a grace period (eg. 5 seconds) from the last sync time on the next sync, to cater for this. But alternatively the issue might bigger than that, which I'm guessing it is. I can't see any indication that the basic delta (or delta for sync targets which support it) retrieves the items / metadata in modified time order, which would suggest that a natural or alphabetical order may be used, and during sync if a new file is created on the target that is ordered before the currently processed item in the list of files in the directory, then it wont get synced (for file system sync the directory listing is not paginated, so I think it reads it into memory and wont receive updated items since the initial listing for sure).

My guess is the problem goes beyond using Joplin with Synthing but can happen with any concurrent editing + syncing during a slow sync operation in Joplin. Take this scenario:
-Client 1 configures sync and does the initial sync to download notes for an existing sync target. Lets say there are 1000 notes on this profile
-During the sync, a change is made for a note which is already connected to the sync target on client 2. This change is synced to the target at the point when client 1 has synced 500 / 1000 notes
-I suspect if the note which was changed on client 2 is within the first 500 notes which already downloaded, then that change will get ignored by client 1 unless a further change is made from client 2.

That is just a hypothesis though and would need to be verified if that is what would actually happen. But if my hypothesis is true, then it still could be hard to solve for every possible scenario, because you wouldn't want to make a grace period too large, as it would increase the amount of items that could be rechecked on every sync. However, you could just not bother catering for scenarios like the initial sync, as we know the initial sync is going to be slow but you could just take a break from using Joplin while you get set up. So maybe a solution could be to make a configurable sync grace period (in seconds) defaulted at 0, but settable up to 5 minutes for example.

EDIT: I tested the above steps and it did not create a problem with some items synced properly, so probably it is an issue more specific to 'mirroring' the sync target on the local filesystem using a tool like Syncthing, which might result in discrepancies with the timestamps on the filesystem

**Comment by mrjo118** [comment]
I've created a PR to partially address this issue, so that new files will never get ignored by the sync. New files are easier to address because there is no need to rely on timestamps for this. Dealing with updates not syncing is a more challenging scenario to tackle.

I posted some analysis at https://discourse.joplinapp.org/t/photos-sometimes-not-syncing-syncthing-android-mac/47083/20?u=mrjo118 which I'm copying here for visibility:

> The main difference between filesystem sync and other types of sync is that with the latter, you've got a single source that is used by all devices, and the whole synchronisation mechanism is built around it. In the case of filesystem sync, there is no single source, but rather the filesystem sync path on each device acts as its own source. When there is just a single source, Joplin can assume that its state is always complete. However, with filesystem sync, there is no such guarantee, and this eventually leads to the issue at hand.

I think this is a very sound analysis of the problem, particularly that the sync is built around the assumption of a single source of truth.

Looking at the basic delta algorithm, it uses a context timestamp to essentially mark that all files with a modified date below this time have been synced. This is ok when you have a single source of truth, because you can assume no new updates can be made on the server with an earlier timestamp to when you last synced. Note that in the sqlite db of the client, we currently store the a sync_time for every item, which is not used for determining the delta, but is used for determining conflicts.

If the sync_time on each item is changed to store the modified date on the server file at the point of synchronising an incoming change (instead of using the current timestamp generated by the client), the algorithm might be able to be modified to cater for multiple sources of truth. The basic delta function currently takes in the ids of all local objects for the client, which is used to compare against a list of all files on the sync target. The input could be modified to take in the sync_time alongside each local object id. Then to handle object updates, instead of comparing against the context timestamp, it could compare against the sync_time value specific for that object. The context timestamp would essentially become redundant, and creation and deletion of objects would be handled as of my open pr, where both are based on the presence of objects on either the client or the server.

All of this can be done without introducing any additional I/O calls to the algorithm, so it should not have any significant impact on performance. This would not affect Joplin server either, which uses a server side database instead of file objects, as it implements its own delta algorothm instead of using the basic delta. The problem then, is this would be a significant change to the syncing algorithm and may not be backwards compatible with existing client and server data. Also, for sync targets which do use basic delta, if the server is not the same device that the client is running on (any sync target which uses basic delta and is not connected to either file system sync or a local webdav server running on the device), then setting sync_time to the modified time of the server file would not be safe, because it could be prone to time sync issues. So this would need to be a behaviour that can be toggled, and potentially a separate column should be used instead of changing the value of sync_time, so that toggle on and off can be done safely.

**Comment by mrjo118** [comment]
I have now found a way to also deal with fixing updates to server objects not getting synced, which involves a toggleable alteration to the basic delta algorithm which is also backwards compatible with the old one, and with existing Joplin clients. So if https://github.com/laurent22/joplin/pull/13054 gets merged, it will fix the issue in it's entirety 😃

**Comment by danielb2** [comment]
awesome work !

**Comment by danielb2** [comment]
🔥🙌
